### PR TITLE
enable dma for USART in stm32f302. added missing hal! implementation

### PR DESCRIPTION
**Pull request by sprhawk**
enable DMA for USARTs for stm32f302, stm32f334
added missing implementation of USART3 for stm32f302, stm32f334
~~added missing implementation of USART4, 5 for stm32f303xb/c/d/e, stm32f302xb/c/d/e, stm32f358, stm32f398~~

### CODE REVIEW

**Comment by sprhawk** [comment]
it can build, but failed to run dma serial transfer on my stm32f302r8 device. I think it should not be caused by changes here. I pushed it for reviewing first

**Comment by strom-und-spiele** [comment]
Thanks for your work!

> it can build, but failed to run dma serial transfer on my stm32f302r8 device ( transferring garbage data). I think it should not be caused by changes here. I pushed it for reviewing first

I don't really get how one can help you here. Do you want a review on the working code or help making the DMA work?

**Comment by ra-kete** [comment]
I also think you'll have an easier time finding people willing to review if you made sure all the CI checks pass first. At least speaking for myself, when I see those failing checks I automatically assume this PR is not worth reviewing yet since I would only find the same issues the CI has already found.

If you have any specific problems we can help you out with, let us know. But as @strom-und-spiele implied, nobody is going to be able to help you much with figuring out what breaks DMA on your F302, since nobody owns that hardware AFAIK.

File: src/dma.rs [hunk 1]
```diff
@@ -738,7 +742,7 @@ macro_rules! on_channel {
     };
 }
 
-#[cfg(feature = "stm32f303")]
+#[cfg(any(feature = "stm32f303", feature = "stm32f302"))]
```
Comment (ra-kete) [inline src/dma.rs:745]: Actually I think this whole `#[cfg]` isn't needed at all. Since we already have the gate in lib.rs.

File: src/lib.rs [hunk 2]
```diff
@@ -116,7 +116,7 @@ pub use crate::pac::interrupt;
 pub mod adc;
 #[cfg(feature = "device-selected")]
 pub mod delay;
-#[cfg(feature = "stm32f303")]
+#[cfg(any(feature = "stm32f303", feature = "stm32f302"))]
```
Comment (ra-kete) [inline src/lib.rs:119]: Nit: 302 should come before 303
(Also applies to the other instances of this in this PR)

**Review by ra-kete — CHANGES_REQUESTED** [review]
Looks reasonable, just a few style requests.

**Review by ra-kete — APPROVED** [review]
LGTM. Can you also add a changelog entry?

**Comment by sprhawk** [comment]
> Looks like you accidentally duplicated this one.

They are different PRs, Should I merge them into one with some text?

**Comment by ra-kete** [comment]
No, just remove the second entry for #132 from the changelog. It's already there (two lines above).

**Comment by sprhawk** [comment]
I have dropped my last commit and forcely pushed

**Comment by ra-kete** [comment]
Hm, now the diff contains changes from other PRs too, that are actually already in master. Not sure what went wrong here. Let me try if I can fix it once I'm back at a PC.

**Comment by sprhawk** [comment]
> Hm, now the diff contains changes from other PRs too, that are actually already in master. Not sure what went wrong here. Let me try if I can fix it once I'm back at a PC.

Because I merge from the upstream, or else, github will not allow merge.

**Comment by ra-kete** [comment]
Sorry, I meant to push my fixup changes to your branch but pushed my master branch instead. This caused an empty diff on this PR, which lead Github to automatically close it. As long as the PR is closed I cannot push to your branch to fix my mistake and as long as the diff is empty I cannot re-open the PR. I took the simple solution of opening a new PR with these changes. Please check out #139. If it looks good to you, I will merge it.

**Comment by sprhawk** [comment]
> Sorry, I meant to push my fixup changes to your branch but pushed my master branch instead. This caused an empty diff on this PR, which lead Github to automatically close it. As long as the PR is closed I cannot push to your branch to fix my mistake and as long as the diff is empty I cannot re-open the PR. I took the simple solution of opening a new PR with these changes. Please check out #139. If it looks good to you, I will merge it.

I think it is fine to me

**Pull request closed without merging**